### PR TITLE
Decouple & move default ports to CliConfiguration

### PR DESCRIPTION
**Pull request by cecton**
Motivation: by decoupling the logic to get the default address from the default port, it will allow the developer to override the ports effortlessly. 

It is particularly relevant in cumulus where we need to override the default ports of the relay chain but we had the re-copy the whole logic as you can see [here](https://github.com/paritytech/cumulus/blob/4c6b959551133bdf2d1f101cb62015e4522ee75d/test/parachain/src/command.rs#L270-L319).

### CODE REVIEW

File: client/cli/src/config.rs [hunk 1]
```diff
@@ -255,17 +255,31 @@ pub trait CliConfiguration: Sized {
...
+	fn rpc_http(&self, _default_port: u16) -> Result<Option<SocketAddr>> {
 		Ok(Default::default())
 	}
 
+	/// Get the default RPC HTTP port.
+	///
+	/// By default this is 9933.
+	fn default_rpc_http_port(&self) -> Result<u16> {
+		Ok(9933)
+	}
```
Comment (expenses) [inline client/cli/src/config.rs:267]: Does this need to return a `Result`?

Comment (cecton) [reply client/cli/src/config.rs:267]: :thinking: the idea with all the results is that it allows the user to put a behavior that can fail.

But for the `_params` method we don't use Result because it makes no sense.

Here it's the first time we use methods to get the default values (the other default values are in `create_configuration` directly).

I don't see why someone would want their default value to fail. But then, is `&self` a useful parameter for a method that returns a default value?

I think it would be best to remove the Result and the `&self` altogether. Let me know if that sounds good/logical to you.

Comment (expenses) [reply client/cli/src/config.rs:267]: I think that removing the `Result` and `&self` is best for now. If a team building on substrate presents a good reason to allow returning an error in this function then we can always allow that later.

Comment (cecton) [reply client/cli/src/config.rs]: Turns out I can't make the variants for the subcommands anymore: https://github.com/paritytech/substrate/pull/6250/commits/c7f985129b956c29ccb1a7bbd6f4e81195f5ef13

I guess that makes sense... :thinking: it wouldn't make much sense to change the default values per subcommand.

I'm starting to wonder if it wouldn't be best to have this kind of default values on the SubstrateCli trait instead of ConfigurationCli.

Comment (expenses) [reply client/cli/src/config.rs]: Oh, well maybe taking `&self` makes sense then.

Comment (cecton) [reply client/cli/src/config.rs]: That kind of default should be valid for a whole program and not customized somehow per subcommand/runcmd. So I think it is better without the `&self` but right now it is badly implemented because it can vary per subcommand. So I'm thinking about moving it to SubstrateCli along the other statics that define what the node is.

The alternative would be to keep the `&self` but not the `Result` but then it means it is still possible to have a different default value per subcommand and also the lack of Result makes it inconsistent with the other methods (maybe rightfully so).

I will move it to SubstrateCli if you don't have any objection. What we can do with them will be more limited but I think it will be less surprising and more consistent overall.

File: client/cli/src/commands/mod.rs [hunk 2]
```diff
@@ -285,15 +285,17 @@ macro_rules! substrate_cli_subcommands {
 				}
 			}
 
-			fn rpc_http(&self) -> $crate::Result<::std::option::Option<::std::net::SocketAddr>> {
+			fn rpc_http(&self, default_port: u16)
+			-> $crate::Result<::std::option::Option<::std::net::SocketAddr>> {
```
Comment (bkchr) [inline client/cli/src/commands/mod.rs:289]:
```suggestion
			-> $crate::Result<std::option::Option<std::net::SocketAddr>> {
```

File: client/cli/src/commands/mod.rs [hunk 3]
```diff
@@ -285,15 +285,17 @@ macro_rules! substrate_cli_subcommands {
...
 				match self {
-					$($enum::$variant(cmd) => cmd.rpc_http()),*
+					$($enum::$variant(cmd) => cmd.rpc_http(default_port)),*
 				}
 			}
 
-			fn rpc_ws(&self) -> $crate::Result<::std::option::Option<::std::net::SocketAddr>> {
+			fn rpc_ws(&self, default_port: u16)
+			-> $crate::Result<::std::option::Option<::std::net::SocketAddr>> {
```
Comment (bkchr) [inline client/cli/src/commands/mod.rs:296]:
```suggestion
			-> $crate::Result<std::option::Option<std::net::SocketAddr>> {
```

File: client/cli/src/lib.rs [hunk 4]
```diff
@@ -210,6 +210,27 @@ pub trait SubstrateCli: Sized {
 		command.init::<Self>()?;
 		Runner::new(self, command)
 	}
+
+	/// Default RPC HTTP port.
+	///
+	/// By default this is 9933.
+	fn default_rpc_http_port() -> u16 {
```
Comment (bkchr) [inline client/cli/src/lib.rs:217]: I don't really understand why he have these ports in this trait here?

Comment (cecton) [reply client/cli/src/lib.rs:217]: I explained in detail here: https://github.com/paritytech/substrate/pull/6250#discussion_r435292134

TL;DR: This is to make sure the default value is the same everywhere (on all subcommands). SubstrateCli has kind of "static"/"constant"-like stuff while ConfigurationCli know how to transform things for a subcommand.

Comment (bkchr) [reply client/cli/src/lib.rs:217]: I'm not entirely sold on this solution. This distinction between these traits seems rather random. And what is with other default values? Why are they not override-able?

Comment (cecton) [reply client/cli/src/lib.rs:217]: Me neither :sweat_smile: initially when I did the refactoring I wanted to have only one trait but it wasn't possible.

Some things work using the parameters provided by the user (CliConfiguration) like: getting the database parameters.

Some other things work using kinda global values (SubstrateCli) like: impl name, impl version.

Some are a bit of both like: getting the chain spec value from user input but process the input the same way for the entire Cli.

In CliConfiguration there are no function that return what should be the default value of something. They are all kinda getter to the user input. Here what we override is the default port of something.

Well, either way I'm out of argument so I will let you decide what is best among those options and I will adapt the code accordingly:

1. Move the default port methods to CliConfiguration and use Result and &self (what I did originally)
2. Move the default port methods to CliConfiguration and not use Result but use &self (what @expenses proposed)
3. Move the default port methods to SubstrateCli and not use Result nor &self (what is it right now)

(Note: there is no option for using CliConfiguration without Result and without &self because &self is required to dispatch to the right subcommand if used in CliConfiguration. This is not the case for SubstrateCli because the trait is kinda global to the entire Cli app.)

cc @seunlanlege (because related to encapsulation OCD)

> The alternative is to have these methods on CliConfiguration and use proc macro magic to override functions

I have been asked to avoid proc macros in the API as much as possible, so this won't even be an option for me.

**Comment by gavofyork** [comment]
anyone else got anything to say about this?

**Comment by bkchr** [comment]
I don't have a better solution nor time to think about something better for now. 

Having support for only overriding 3 default values is also somewhat weird. Supporting more brings us probably into the same situation as with the service. 

But with having `SubstrateCli` starting to expose default values and `CliConfiguration` for certain values having not changeable default values is weird.

**Comment by cecton** [comment]
Ok I will move it back and we merge?

**Comment by cecton** [comment]
Ok I will look into it

**Comment by bkchr** [comment]
Ty

**Comment by cecton** [comment]
Closing because #6857 has been merged
